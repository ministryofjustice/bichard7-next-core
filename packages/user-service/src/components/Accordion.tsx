import { ReactNode } from "react"

interface Props {
  children: ReactNode
}

const Accordion = ({ children }: Props) => (
  <div className="govuk-accordion" data-module="govuk-accordion">
    {children}
  </div>
)

export default Accordion
