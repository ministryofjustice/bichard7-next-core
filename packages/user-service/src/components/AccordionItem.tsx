import { ReactNode } from "react"

interface Props {
  children: ReactNode
  dataTest?: string
  heading: string
  id: string
}

const AccordionItem = ({ children, dataTest, heading, id }: Props) => (
  <>
    <div className="govuk-accordion__section" data-test={dataTest}>
      <div className="govuk-accordion__section-header">
        <h2 className="govuk-accordion__section-heading">
          <span className="govuk-accordion__section-button" id={id}>
            {heading}
          </span>
        </h2>
      </div>
      <div className="govuk-accordion__section-content" aria-labelledby={id}>
        {children}
      </div>
    </div>
  </>
)

export default AccordionItem
