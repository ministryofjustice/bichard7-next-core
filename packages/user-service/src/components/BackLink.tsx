import Link from "./Link"

interface Props {
  href: string
}

const BackLink = ({ href }: Props) => (
  <Link href={href} className="govuk-back-link" data-test="back-link">
    {"Back"}
  </Link>
)

export default BackLink
