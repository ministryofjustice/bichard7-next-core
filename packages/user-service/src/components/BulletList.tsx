import React from "react"

interface BulletListProps {
  heading?: string
  items: string[]
  listClassName?: string
  headingClassName?: string
}

const BulletList: React.FC<BulletListProps> = ({
  heading,
  items,
  listClassName = "govuk-list govuk-list--bullet",
  headingClassName = "govuk-body"
}) => {
  if (!items || items.length === 0) {
    return null
  }

  return (
    <>
      {heading && <p className={headingClassName}>{heading}</p>}

      <ul className={listClassName}>
        {items.map((item, index) => (
          <li key={index}>{item}</li>
        ))}
      </ul>
    </>
  )
}

export default BulletList
