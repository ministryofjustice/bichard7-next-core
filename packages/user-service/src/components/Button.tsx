import { ReactNode } from "react"
import classnames from "classnames"

type ButtonVariant = "default" | "warning" | "secondary"
interface Props {
  children: ReactNode
  variant?: ButtonVariant
  noDoubleClick?: boolean
  id?: string
  className?: string
  name?: string
  value?: string
  isDisabled?: boolean
  dataTest?: string
}

const getButtonClass = (variant?: ButtonVariant): string => {
  const variantClass = !variant || variant === "default" ? "" : ` govuk-button--${variant}`
  return `govuk-button${variantClass}`
}

const Button = ({ children, variant, noDoubleClick, id, className, name, value, isDisabled, dataTest }: Props) => {
  const classes = classnames(getButtonClass(variant), className && { [className]: !!className })

  return (
    <button
      type="submit"
      className={classes}
      name={name}
      value={value}
      data-module="govuk-button"
      data-prevent-double-click={noDoubleClick}
      data-test={dataTest}
      disabled={isDisabled}
      id={id}
    >
      {children}
    </button>
  )
}

export default Button
