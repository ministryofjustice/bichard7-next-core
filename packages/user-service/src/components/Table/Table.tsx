import { ReactNode } from "react"
import KeyValuePair from "types/KeyValuePair"
import getColumnComponents from "./getColumnComponents"
import TableColumn from "./TableColumn"

export type TableHeader = [string, string]

export type TableHeaders = Array<TableHeader>

type Props = {
  tableHeaders: TableHeaders
  tableTitle?: string
  tableData: KeyValuePair<string, string>[]
  children?: ReactNode
}

const Table = ({ tableTitle, tableHeaders, tableData, children }: Props) => {
  const columnComponents = getColumnComponents(children)

  return (
    <table className="govuk-table">
      {tableTitle && <caption className="govuk-table__caption govuk-table__caption--m">{tableTitle}</caption>}
      <thead className="govuk-table__head">
        <tr className="govuk-table__row">
          {tableHeaders.map((header: TableHeader) => (
            <th key={header[0]} scope="col" className="govuk-table__header">
              {header[1]}
            </th>
          ))}
        </tr>
      </thead>
      <tbody className="govuk-table__body">
        {tableData.map((row: KeyValuePair<string, string>) => (
          <tr key={Object.values(row).join("")} className="govuk-table__row">
            {tableHeaders.map((header: TableHeader) => {
              const field = header[0]
              return <TableColumn key={field} component={columnComponents[field]} item={row} field={field} />
            })}
          </tr>
        ))}
      </tbody>
    </table>
  )
}

export default Table
