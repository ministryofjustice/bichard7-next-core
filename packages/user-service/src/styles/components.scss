@import "variables";

/* B7 Inset Text */
.b7-inset-text {
  display: block;
  font-weight: 300;
  font-size: 1rem;
  padding: 0.9375rem 1.25rem;
  background: var(--other-light-grey, $b7-light-grey-color);
  box-shadow: 5px 0px 0px 0px $b7-dark-grey-color inset;

  .b7-inset-text__heading {
    display: block;
    font-weight: 700;
  }
}

/* B7 Accordion */
.b7-accordion {
  @extend .govuk-accordion__section;

  .b7-accordion__content {
    margin-top: 0.75rem;
    margin-bottom: 0.5rem;
  }

  .b7-accordion__header {
    @extend .govuk-accordion__section-header;
    margin-bottom: 0.5rem;

    .b7-accordion__heading {
      @extend .govuk-accordion__section-heading;

      .b7-accordion__button {
        @extend .govuk-accordion__section-button;

        .b7-accordion__toggle {
          @extend .govuk-accordion__section-toggle;

          .b7-accordion__toggle-focus {
            @extend .govuk-accordion__section-toggle-focus;

            .b7-accordion-nav__chevron {
              @extend .govuk-accordion-nav__chevron;
            }

            .b7-accordion-nav__chevron--down {
              @extend .govuk-accordion-nav__chevron--down;
            }

            .b7-accordion__toggle-text {
              @extend .govuk-accordion__section-toggle-text;
            }
          }
        }
      }
    }
  }
}

.js-enabled .b7-accordion .b7-accordion__header .b7-accordion__heading .b7-accordion__button,
.b7-accordion .b7-accordion__header .b7-accordion__heading .js-enabled .b7-accordion__button {
  padding: 0;
  border: none;
}

.js-enabled .b7-accordion .b7-accordion__header .b7-accordion__heading .b7-accordion__button .b7-accordion__toggle,
.b7-accordion .b7-accordion__header .b7-accordion__heading .b7-accordion__button .js-enabled .b7-accordion__toggle {
  margin-bottom: 0;
}

/* MoJ badge */
.moj-badge {
  background-color: white;
}

.preview-button {
  margin-Bottom: 0px;
  white-space: nowrap;
}