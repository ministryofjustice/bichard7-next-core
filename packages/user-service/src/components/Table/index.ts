import Table from "./Table"
import LinkColumn from "./LinkColumn"
import type { TableHeader, TableHeaders } from "./Table"

export { Table, LinkColumn }
export type { TableHeader, TableHeaders }
