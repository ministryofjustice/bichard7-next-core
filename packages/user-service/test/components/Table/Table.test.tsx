import { render } from "@testing-library/react"
import { Table, TableHeaders } from "components/Table"
import KeyValuePair from "types/KeyValuePair"

const tableData: KeyValuePair<string, string>[] = [{ column1: "Column 1 value" }, { column2: "Column 2 value" }]

const tableHeaders: TableHeaders = [
  ["column1", "First Column"],
  ["column2", "Second Column"]
]

it("should render text column when no specific column type provided", () => {
  const { container } = render(<Table tableHeaders={tableHeaders} tableTitle="Testing Table" tableData={tableData} />)

  expect(container).toMatchSnapshot()
})

it("should render custom column when custom column field name exists", () => {
  const CustomColumn = ({ field }: { field: string }) => <div>{`Custom column for field ${field}`}</div>

  const { container } = render(
    <Table tableHeaders={tableHeaders} tableTitle="Testing Table" tableData={tableData}>
      <CustomColumn field="column1" />
    </Table>
  )

  expect(container).toMatchSnapshot()
})

it("should not render custom column when custom column field name does not exist", () => {
  const CustomColumn = ({ field }: { field: string }) => <div>{`Custom column for field ${field}`}</div>

  const { container } = render(
    <Table tableHeaders={tableHeaders} tableTitle="Testing Table" tableData={tableData}>
      <CustomColumn field="invalidFieldName" />
    </Table>
  )

  expect(container).toMatchSnapshot()
})
