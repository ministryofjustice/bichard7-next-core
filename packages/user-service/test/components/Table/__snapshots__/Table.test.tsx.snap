// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`should not render custom column when custom column field name does not exist 1`] = `
<div>
  <table
    class="govuk-table"
  >
    <caption
      class="govuk-table__caption govuk-table__caption--m"
    >
      Testing Table
    </caption>
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          First Column
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Second Column
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          Column 1 value
        </td>
        <td
          class="govuk-table__cell"
        />
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        />
        <td
          class="govuk-table__cell"
        >
          Column 2 value
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`should render custom column when custom column field name exists 1`] = `
<div>
  <table
    class="govuk-table"
  >
    <caption
      class="govuk-table__caption govuk-table__caption--m"
    >
      Testing Table
    </caption>
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          First Column
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Second Column
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          <div>
            Custom column for field column1
          </div>
        </td>
        <td
          class="govuk-table__cell"
        />
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          <div>
            Custom column for field column1
          </div>
        </td>
        <td
          class="govuk-table__cell"
        >
          Column 2 value
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`should render text column when no specific column type provided 1`] = `
<div>
  <table
    class="govuk-table"
  >
    <caption
      class="govuk-table__caption govuk-table__caption--m"
    >
      Testing Table
    </caption>
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          First Column
        </th>
        <th
          class="govuk-table__header"
          scope="col"
        >
          Second Column
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          Column 1 value
        </td>
        <td
          class="govuk-table__cell"
        />
      </tr>
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        />
        <td
          class="govuk-table__cell"
        >
          Column 2 value
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
